frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/index.ts

```diff
@@ -42,25 +42,56 @@ function getWaterfallExtent(dataset: WaterfallDataset) {
   return extent;
 }
 
-function getSortedRows(
+function getSortedAggregatedRows(
   rows: RowValues[],
   cardColumns: CartesianChartColumns,
   settings: ComputedVisualizationSettings,
 ) {
-  if (settings["graph.x_axis.scale"] !== "timeseries") {
-    return rows;
+  if (settings["graph.x_axis.scale"] === "timeseries") {
+    rows.sort((left, right) => {
+      const leftValue = left[cardColumns.dimension.index];
+      const rightValue = right[cardColumns.dimension.index];
+
+      if (typeof leftValue === "string" && typeof rightValue === "string") {
+        return dayjs(leftValue).valueOf() - dayjs(rightValue).valueOf();
+      }
+
+      return 0;
+    });
   }
 
-  return rows.sort((left, right) => {
-    const leftValue = left[cardColumns.dimension.index];
-    const rightValue = right[cardColumns.dimension.index];
+  // aggregate metric values by dimension into a map
+  const columns = assertMultiMetricColumns(cardColumns);
+  const dimensionMetricMap = new Map<RowValue, number>();
+
+  rows.forEach(row => {
```

Comment:
For tooltips we need to aggregate all metrics, even the ones that are not being used for the waterfall series.
I think you can do it in one run here:
- have aggregated rows map `Map<RowValue, RowValue[]>`: dimension value to row
- iterate through rows
    - if aggregated rows do not have the dimension value set the current row
    - if aggregated rows already have a row then aggregate all values
   
Also, we do not consider nulls as 0, they are rendered as `(empty)` in tooltips. Maybe you could reuse the `sumMetric` function from `frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts`

Reply:
Can we maybe reuse `getJoinedCardsDataset` to have exactly the same behavior? The shape is different, to access values you will need to create dataKeys for dimension and metric but it should not be hard.

But it would be beneficial because if you expose this raw dataset as `dataset` property which matches the structure of combo and scatter charts, tooltips and click behavior will work right away because the data format is the same which would reduce duplication

Reply:
Sure, I'll make these changes when I implement the dynamic chart since it'll be easier to test with the actual tooltip. For now can you approve this PR?

Reply:
@EmmadUsmani it still has the issue I mentioned about nulls are being considered as 0 by this code `const newMetric = currMetric + (existingMetric ?? 0);`. Please don't forget about this issue

Reply:
Sure, I've filed an issue for it: https://github.com/metabase/metabase/issues/37288